addons/mail/static/src/rtc/voice_recorder.js

```diff
@@ -0,0 +1,96 @@
+/* @odoo-module */
+import { Component, useState, onWillUnmount } from "@odoo/owl";
+import { useService } from "@web/core/utils/hooks";
+import { sprintf } from "@web/core/utils/strings";
+import { _t } from "@web/core/l10n/translation";
+import { browser } from "@web/core/browser/browser";
+
+export class VoiceRecorder extends Component {
+    static props = ["attachmentUploader"];
+    static template = "mail.VoiceRecorder";
+
+    microphone = null;
+    config = null;
+    recorder = null;
+    blob = null;
+
+    setup() {
+        this.state = useState({ recording: false });
+        this.notification = useService("notification");
+        this.userSettings = useService("mail.user_settings");
+        this.isSafari = /^((?!chrome|android).)*safari/i.test(navigator.userAgent);
+        onWillUnmount(() => {
+            this.unmount = true;
+            this.stopRecording();
+        });
+    }
+
+    async startRecording() {
+        if (!this.microphone) {
+            try {
+                this.microphone = await browser.navigator.mediaDevices.getUserMedia({
+                    audio: this.userSettings.audioConstraints,
+                });
+            } catch {
+                this.notification.add(
+                    sprintf(_t('"%(hostname)s" needs to access your microphone'), {
+                        hostname: window.location.host,
+                    }),
+                    { type: "warning" }
+                );
+                return;
+            }
+        }
+        this.state.recording = true;
+        const config = {
+            type: "audio",
+            mimeType: this.isSafari ? "audio/mp4" : "audio/webm",
+        };
+
+        this.recorder = new MediaRecorder(this.microphone, config);
+        this.recorder.ondataavailable = (e) => {
+            if (this.unmount) {
+                return;
+            }
+            this.blob = config.getNativeBlob
+                ? e.data
+                : new Blob([e.data], {
+                      type: this.recorder.mimeType || config.mimeType,
+                  });
+            const file = new File([this.blob], this.filename(), {
+                type: "audio/mp3",
+            });
+            this.props.attachmentUploader.uploadFile(file);
+            this.recorder = null;
+        };
+        this.recorder.start(24 * 60 * 60 * 1000);
+    }
+
+    stopRecording() {
+        if (this.recorder?.state === "recording") {
+            this.recorder?.stop();
```

Comment:
You don't need to put the conditional chaining given the above condition,
I suggested it because it becomes necessary if you check `state !== "inactive"` to cover the `paused` state.